### PR TITLE
Implement geometry volumes for fiber grids in BEMC ScFi

### PR DESCRIPTION
**Pull request by Chao1009**
### Briefly, what does this PR introduce?
It adds geometry for grids so we can retrieve the position/dimension of a grid during the digitization/reconstruction.

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [x] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [x] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
No, it updates the internal behavior of geometry construction for the BarrelCalorimeterInterlayers. It does not change any of the interfaces to use it.

### Does this PR change default behavior?
No.

### CODE REVIEW

**Comment by Chao1009** [comment]
Working on a visualization script for showing grid/fibers on local X-Z plane

**Comment by Chao1009** [comment]
It's a part of the solution to https://github.com/eic/EICrecon/issues/499

File: src/BarrelCalorimeterInterlayers_geo.cpp [hunk 1]
```diff
@@ -192,71 +216,65 @@ void buildFibers(Detector& desc, SensitiveDetector& sens, Volume& s_vol, xml_com
...
+  for (auto &gr : grids) {
+    Assembly grid_vol(Form("fiber_grid_%i_%i", gr.ix, gr.iz));
+    // fiber is along y-axis of the layer volume, so grids are arranged on X-Z plane
+    Transform3D gr_tr(RotationZYX(0, 0, M_PI * 0.5), Position(gr.mean_centroid.x(), 0, gr.mean_centroid.y()));
+    auto grid_phv = s_vol.placeVolume(grid_vol, gr_tr);
+    grid_phv.addPhysVolID(f_id_grid, gr.ix + gr.iz * grid_div.first + 1);
+
+    // loop over all fibers that are not assigned to a grid
+    int f_id = 1;
+    for (auto &fi : fibers | boost::adaptors::filtered( [](const Fiber &f) {return not f.assigned;} )) {
```
Comment (wdconinc) [inline src/BarrelCalorimeterInterlayers_geo.cpp:255]: No c++20 features for now. We can't use c++20 until ROOT can be installed with c++20 support.

**Review by wdconinc — COMMENTED** [review]
Some comments, more later.

File: src/BarrelCalorimeterInterlayers_geo.cpp (hunk 1 above)
Comment (Chao1009) [reply src/BarrelCalorimeterInterlayers_geo.cpp:255]: OK, I'll change these

File: src/BarrelCalorimeterInterlayers_geo.cpp [hunk 2]
```diff
@@ -18,17 +18,41 @@
...
 using namespace std;
 using namespace dd4hep;
 using namespace dd4hep::detail;
 
 typedef ROOT::Math::XYPoint Point;
 // fiber placement helpers, defined below
-vector<vector<Point>> fiberPositions(double radius, double x_spacing, double z_spacing, double x, double z, double phi,
-                                     double spacing_tol = 1e-2);
+struct FiberGrid {
+  int ix = 0, iz = 0;
```
[the remark this answers is not in the document]
Comment (Chao1009) [reply src/BarrelCalorimeterInterlayers_geo.cpp]: Fixed in Fixed in [84c16547](https://github.com/eic/epic/pull/370/files/dd54011b4a072dc386e77393aac1e2b4bffdc1e1..006c43a2ea6cae1925ee60ffd2c25fe084c16547)

File: src/BarrelCalorimeterInterlayers_geo.cpp [hunk 3]
```diff
@@ -18,17 +18,41 @@
 #include "Math/Point2D.h"
 #include "TGeoPolygon.h"
 #include "XML/Layering.h"
+#include <functional>
+#include <boost/range/adaptor/filtered.hpp>
 
 using namespace std;
 using namespace dd4hep;
 using namespace dd4hep::detail;
```
[the remark this answers is not in the document]
Comment (Chao1009) [reply src/BarrelCalorimeterInterlayers_geo.cpp]: Fixed in [84c16547](https://github.com/eic/epic/pull/370/files/dd54011b4a072dc386e77393aac1e2b4bffdc1e1..006c43a2ea6cae1925ee60ffd2c25fe084c16547)

**Comment by Chao1009** [comment]
The geometry visualization script is done
grid_fibers

**Comment by Chao1009** [comment]
The same type of plot with different stave and layer (check title)
grid_fibers

**Comment by Chao1009** [comment]
Add more adjacent la
grid_fibers
yers/grids in the visualization script to check the geometry.

**Comment by Chao1009** [comment]
This PR is ready.

**Comment by Chao1009** [comment]
Caught a minor issue with fiber alignment between adjacent layers. Fixed as
image

File: scripts/subdetector_tests/draw_bemc_scfi_grids.py [hunk 4]
```diff
@@ -1,5 +1,8 @@
 '''
```
Comment (wdconinc) [inline scripts/subdetector_tests/draw_bemc_scfi_grids.py:1]: Can you add an SPDX header here to pass the checks?

**Review by wdconinc — CHANGES_REQUESTED** [review]

File: scripts/subdetector_tests/draw_bemc_scfi_grids.py (hunk 4 above)
Comment (Chao1009) [reply scripts/subdetector_tests/draw_bemc_scfi_grids.py:3]: Add it to the commit https://github.com/eic/epic/pull/370/commits/38c228c9f9f239721080466018646a979781dc43

File: scripts/subdetector_tests/draw_bemc_scfi_grids.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
+# SPDX-License-Identifier: LGPL-3.0-or-later
+# Copyright (C) 2022 Chao Peng
```
Comment (wdconinc) [inline scripts/subdetector_tests/draw_bemc_scfi_grids.py:2]:
```suggestion
# Copyright (C) 2023 Chao Peng
```

**Review by wdconinc — APPROVED** [review]

**Pull request merged**